p:fileUpload: Exception if one or more filenames has a '%' character

This is the exception: 
javax.servlet.ServletException: URLDecoder: Illegal hex characters in escape (%) pattern - For input string: ".j"

This is the stacktrace:

Caused by: java.lang.IllegalArgumentException: URLDecoder: Illegal hex characters in escape (%) pattern - For input string: ".j"
at java.net.URLDecoder.decode(URLDecoder.java:194)
at org.primefaces.model.NativeUploadedFile.decode(NativeUploadedFile.java:180)
at org.primefaces.model.NativeUploadedFile.getContentDispositionFileName(NativeUploadedFile.java:156)
at org.primefaces.model.NativeUploadedFile.resolveFilename(NativeUploadedFile.java:114)
at org.primefaces.model.NativeUploadedFile.(NativeUploadedFile.java:45)
at org.primefaces.component.fileupload.NativeFileUploadDecoder.decodeAdvanced(NativeFileUploadDecoder.java:67)
at org.primefaces.component.fileupload.NativeFileUploadDecoder.decode(NativeFileUploadDecoder.java:38)
at org.primefaces.component.fileupload.FileUploadRenderer.decode(FileUploadRenderer.java:62)

- Feel free to provide a PR (Primefaces is an open-source project, any fixes or improvements are welcome.)

I found this discussion on stackoverflow that seem related to this behaviour:

https://stackoverflow.com/questions/34633956/any-concerns-about-executing-urldecoder-against-a-url-that-was-not-encoded


1) Environment
- PrimeFaces version: v.6.2.7

- Does it work on the newest released PrimeFaces version? Version?

- Application server + version:

jsf Mojarra 2.2.14
Tomcat 8.5
Java 8u181

- Affected browsers: All

2) Expected behavior

Upload the file or remove the '%' character from filename 

3) Actual behavior

Exception

4) Steps to reproduce

Go to:  https://www.primefaces.org/showcase/ui/file/upload/single.xhtml

Try to upload a jpg file with this filename:   test%.jpg

--- a/src/main/java/org/primefaces/model/NativeUploadedFile.java
+++ b/src/main/java/org/primefaces/model/NativeUploadedFile.java
@@ -177,6 +177,9 @@ public class NativeUploadedFile implements UploadedFile, Serializable {
 
     private String decode(String encoded) {
         try {
+            // GitHub #3916 escape + and % before decode
+            encoded = encoded.replaceAll("%(?![0-9a-fA-F]{2})", "%25");
+            encoded = encoded.replaceAll("\\+", "%2B");
             return URLDecoder.decode(encoded, "UTF-8");
         }
         catch (UnsupportedEncodingException ex) {

--- a/src/test/java/org/primefaces/model/NativeUploadedFileTest.java
+++ b/src/test/java/org/primefaces/model/NativeUploadedFileTest.java
@@ -61,6 +61,30 @@ public class NativeUploadedFileTest {
         // Assert
         Assert.assertEquals("Test;123.txt", output);
     }
+    
+    @Test
+    public void testPercent() {
+        // Arrange
+        final String input = "form-data; name=\"XXX:XXX\"; filename=\"test%.jpg\"; charset=\"UTF-8\"";
+
+        // Act
+        final String output = file.getContentDispositionFileName(input);
+
+        // Assert
+        Assert.assertEquals("test%.jpg", output);
+    }
+    
+    @Test
+    public void testPlus() {
+        // Arrange
+        final String input = "form-data; name=\"XXX:XXX\"; filename=\"test+.jpg\"; charset=\"UTF-8\"";
+
+        // Act
+        final String output = file.getContentDispositionFileName(input);
+
+        // Assert
+        Assert.assertEquals("test+.jpg", output);
+    }
 
     @Test
     public void testSlashes() {

